lib/cloud_controller/metrics/periodic_updater.rb

```diff
@@ -75,10 +78,12 @@ def update_job_queue_length
 
       total                      = 0
       pending_job_count_by_queue = jobs_by_queue_with_count.each_with_object({}) do |row, hash|
+        @known_job_queues[row[:queue].to_sym] = 0
         total += row[:count]
         hash[row[:queue].to_sym] = row[:count]
       end
 
+      pending_job_count_by_queue = @known_job_queues.merge(pending_job_count_by_queue)
```

Comment:
the name makes this opaque to me... before doing this merge pending_job_count_by_queue is different how? what do we accomplish by merging them?

Reply:
@known_job_queues is building up a persistent hash of queues that have been seen before with 0 as a default value since if the queue isn't returned by the database query we know the count is 0.

This ends up looking like:

`{"some-queue" => 0, "some-other-queue" => 0}`

The pending_job_count_by_queue hash only contains things that have non-zero counts. It ends up looking something like:

`{"some-other-queue" => 1}`

The merge operation will merge the two and return a new hash. The values from the argument hash take precedent.

So it becomes `{"some-queue" => 0, "some-other-queue" => 1}`

I realize though that since we're using Rails / ActiveSupport I could use `reverse_merge!` (https://apidock.com/rails/v5.2.3/Hash/reverse_merge%21) which will make the calling hash's values take precedence. This might make this look more intuitive.

So instead we'd have like:

```ruby
pending_job_count_by_queue.reverse_merge!(@known_job_queues)
```

There are also more verbose ways of doing this. This just felt more Rubyish.

Reply:
Here's the Set approach:

```ruby
    def update_job_queue_length
      jobs_by_queue_with_count = Delayed::Job.where(attempts: 0).group_and_count(:queue)

      total                      = 0
      pending_job_count_by_queue = jobs_by_queue_with_count.each_with_object({}) do |row, hash|
        @known_job_queues.add(row[:queue].to_sym)
        total += row[:count]
        hash[row[:queue].to_sym] = row[:count]
      end

      @known_job_queues.each do |queue|
        pending_job_count_by_queue[queue] ||= 0
      end

      @updaters.each { |u| u.update_job_queue_length(pending_job_count_by_queue, total) }
    end
```

Reply:
`reverse_merge!` is best IMO, it disambiguates the structure of of the final hash and shows that the transformation is being done in-place.

Reply:
Switched to the `reverse_merge!` version:
https://github.com/cloudfoundry/cloud_controller_ng/commit/0fddfd413e89f5c14fe9f71241628e7aeadb68c7#diff-e263d41f0b0a3927a046dcd948d895f6R106